ddtrace/internal/datadog/profiling/dd_wrapper/include/sample_manager.hpp

```diff
@@ -0,0 +1,254 @@
+#pragma once
+
+#include "sample.hpp"
+#include "types.hpp"
+
+#include <array>
+#include <atomic>
+#include <optional>
+#include <vector>
+
+namespace Datadog {
+
+// Tradeoff: either Invalid is 0 (nice: uninitialized state is invalid) or
+//           it's >_Length (nice: can place real sample types in an array + index by enum)
+// Choosing the latter for simplicity and because we own all downstream consumption for now
+enum class SampleHandle
+{
+    Stack,
+    Lock,
+    Allocation,
+    Heap,
+    _Length,
+    Invalid,
+};
+
+using SampleStorage = std::vector<Sample>;
+
+class SampleManager
+{
+  private:
+    static inline std::array<std::atomic<bool>, static_cast<size_t>(SampleHandle::_Length)> handler_state{};
```

Comment:
Do we want everything `static`

Reply:
I think we've reduced this a lot in the latest iteration, but we can do better.  It'd require some more refactors, but I think I'd like to tack that on as a followup PR.  For discussion, consider the following

UploaderBuilder has a lot of `static`, but it is actually treated as a singleton.  It doesn't have to be either.  Of its 10 static objects, 9 can be factored into a context class.  The remainder is a lock which we may have to keep.

The 3 statics in SampleManager are the exact same story.  Two can be factored into a single context object, for instance, with the remainder being a lock.

I'm not 100% sure how to factor out the static-ness of the lock or cancellation token in the uploader, nor the static Profile in the Sample, without significant refactors.

Either way, TBD.